### PR TITLE
Add ad extensions reporter script

### PR DESCRIPTION
**Pull request by gideon-brainlabs**
The script takes all the extensions in an account, and outputs them to a GSheet.

There is a separate tab for each extension type, containing the details of all the extensions, as well as an overall tab with the counts of feed items with each extension type.

### CODE REVIEW

File: Reporting/ad-extensions-reporter.js [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+
+var spreadsheetUrl = "https://docs.google.com/YOUR-SPREADSHEET-URL-HERE";
+// The URL of the Google Doc the results will be put into.
+
+var includeExtensionsWithZeroImpressions = true;
+// Keep as true if you also want to see the extensions that aren't being shown. This is the default,
+// and recommended if you are using this script to get an overview of all the extensions in your account.
+// In accounts with a very large number of extensions, this script is too slow. If that is the case
+// for your account, you can change this variable's value to false to speed it up. That will exclude
+// extensions with no impressions today.
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: Just cut down a few words to make it shorter, good comment though!
```suggestion
// Keep as true if you also want to see the extensions that aren't being shown. This is the default,
// and recommended if you are using this script to get an overview of all the extensions in your account.
// (In accounts with a very large number of extensions, this script is too slow, so you can change this to 
// false to speed it up. That will exclude extensions with no impressions today.)
```

File: Reporting/ad-extensions-reporter.js [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+////////////////////////////////////////////////////////////////////////////////
+// Functions
+
+function main() {
+    var spreadsheet = checkSpreadsheet(spreadsheetUrl, "Output Spreadsheet");
+    var sheets = spreadsheet.getSheets();
+
+    Logger.log("Preparing spreadsheet");
+    try {
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: Sheets calls can error for a number of reasons (like google servers being temporarily unavailable). While these are going to be rare compared to the error you are expecting, which is that the sheet doesn't exist, this kind of model, where you try catch expecting a specific kind of error, and then executing code _assuming_ that the error you got was the error you were expecting, can sometimes end up being wrong. (e.g. if it was a random google error, and the sheet does already exist, then this code will try and make the sheet again, and then error on that!)

If you want to do an ask for forgiveness rather than permission model, then you'd want to check that the error you got explicitly matches the one you are expecting to indicate the lack of a sheet, and otherwise throw the exception.

^ This applies for any code you are writing in the future too - always make sure the exception is the one you were expecting!

In this case, though I'd probably recommend, checking if the sheet exists first:
```
var allExtensionsSheet = spreadsheet.getSheetByName("All Extensions");
if(allExtensionsSheet === null) {
    allExtensionsSheet = spreadsheet.insertSheet("All Extensions");
}
allExtensionsSheet.clear();
```

File: Reporting/ad-extensions-reporter.js [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+            if (Object.keys(duplicateAttributeValuesCounts).indexOf(jsonAttributeValues) === -1) {
+                duplicateAttributeValuesCounts[jsonAttributeValues] = 2;
+            } else {
+                duplicateAttributeValuesCounts[jsonAttributeValues] += 1;
+            };
+            for (var i = 0, n = disapprovalShortNames.length; i < n; i++) {
+                var disapprovalName = disapprovalShortNames[i];
+                if (disapprovalShortNamesByExtensionType[placeholderTypeId][
+                    jsonAttributeValuesByExtensionType[placeholderTypeId].indexOf(jsonAttributeValues)
+                ].indexOf(disapprovalName) === -1) {
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: This is quite a complicated concept, and quite long - I think this would benefit from a variable name - shortens the line length, and lets you include some text that explains what is happening

File: Reporting/ad-extensions-reporter.js [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+                    disapprovalShortNamesByExtensionType[placeholderTypeId][
+                        jsonAttributeValuesByExtensionType[placeholderTypeId].indexOf(jsonAttributeValues)
+                    ] = disapprovalShortNamesByExtensionType[placeholderTypeId][
+                        jsonAttributeValuesByExtensionType[placeholderTypeId].indexOf(jsonAttributeValues)
+                    ].concat(disapprovalName);
+                }
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]:
```suggestion
            for (var i = 0, n = disapprovalShortNames.length; i < n; i++) {
                var disapprovalName = disapprovalShortNames[i];
                var previouslyFoundDisapprovalNames = disapprovalShortNamesByExtensionType[placeholderTypeId][
                    jsonAttributeValuesByExtensionType[placeholderTypeId].indexOf(jsonAttributeValues)
                ];
                if (previouslyFoundDisapprovalNames.indexOf(disapprovalName) === -1) {
                    disapprovalShortNamesByExtensionType[placeholderTypeId][
                        jsonAttributeValuesByExtensionType[placeholderTypeId].indexOf(jsonAttributeValues)
                    ] = previouslyFoundDisapprovalNames.concat(disapprovalName);
                }
```

File: Reporting/ad-extensions-reporter.js (hunk 3 above)
Comment (emma-brainlabs) [reply Reporting/ad-extensions-reporter.js]: e.g. see below

File: Reporting/ad-extensions-reporter.js [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+        }
+        extensionTypeFeedItemCounts[placeholderTypeId] += 1;
+    }
+
+    // Create a spreadsheet tab for each extension type, and output its data
+    Logger.log("Outputting data to spreadsheet");
+    for (var placeholderTypeId in jsonAttributeValuesByExtensionType) {
+        var placeholderTable = extensionPlaceholderTypeIdToFieldIdsObjectMapping[placeholderTypeId];
+        var orderedAttributeIds = Object.keys(placeholderTable).sort();
+        orderedAttributeIds.pop();
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: Does this remove the name? maybe worth adding in a comment saying that is what you are doing? As without remembering the name is the first element, and isn't relevant for the rest of this code, you might not know!

File: Reporting/ad-extensions-reporter.js [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+        var jsonAttributeValues = row['AttributeValues'];
+        var jsonDisapprovalShortNames = row['DisapprovalShortNames'];
+        if ('' === jsonDisapprovalShortNames.toString()) {
+            var disapprovalShortNames = [];
+        } else {
+            var disapprovalShortNames = JSON.parse(jsonDisapprovalShortNames);
+        };
+        if (jsonAttributeValuesByExtensionType[placeholderTypeId].indexOf(jsonAttributeValues) === -1) {
+            jsonAttributeValuesByExtensionType[placeholderTypeId].push(jsonAttributeValues);
+            disapprovalShortNamesByExtensionType[placeholderTypeId].push(disapprovalShortNames);
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: Here you could add `duplicateAttributeValuesCounts[jsonAttributeValues] = 1`, then you wouldn't need the 2 check on like 282, or the 1 check on line 322

File: Reporting/ad-extensions-reporter.js [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+                    } else {
+                        return '';
+                    }
+                }
+            );
+            rowValues.push(feedItemCount, disapprovalShortNames.join(','));
+            sheetOutput.push(rowValues);
+        }
+
+        sheetOutput.sort(function (rowValuesA, rowValuesB) {
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js:276]: What does this sort on/ what does this do? Bit confused!

File: Reporting/ad-extensions-reporter.js [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+                ].indexOf(disapprovalName) === -1) {
+                    disapprovalShortNamesByExtensionType[placeholderTypeId][
+                        jsonAttributeValuesByExtensionType[placeholderTypeId].indexOf(jsonAttributeValues)
+                    ] = disapprovalShortNamesByExtensionType[placeholderTypeId][
+                        jsonAttributeValuesByExtensionType[placeholderTypeId].indexOf(jsonAttributeValues)
+                    ].concat(disapprovalName);
+                }
+            }
+        }
+        extensionTypeFeedItemCounts[placeholderTypeId] += 1;
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: Could probably eliminate this variable and then work it out at the end, by using `jsonAttributeValuesByExtensionType`?

File: Reporting/ad-extensions-reporter.js [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+
+        var jsonAttributeValues = row['AttributeValues'];
+        var jsonDisapprovalShortNames = row['DisapprovalShortNames'];
+        if ('' === jsonDisapprovalShortNames.toString()) {
+            var disapprovalShortNames = [];
+        } else {
+            var disapprovalShortNames = JSON.parse(jsonDisapprovalShortNames);
+        };
+        if (jsonAttributeValuesByExtensionType[placeholderTypeId].indexOf(jsonAttributeValues) === -1) {
+            jsonAttributeValuesByExtensionType[placeholderTypeId].push(jsonAttributeValues);
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: As this data is linked together (as in you have one jsonAttributeValues that maps to an array of disapprovalShortNames), rather than having two arrays, each containing one, and being intrinsically linked (aka relying on the fact that you always add to them at the same time, so they are always in the same order), you could have one thing e.g. `extensions`
```javascript
if (Object.keys(extensions[placeholderTypeId]).indexOf(jsonAttributeValues) === -1) {
            extensions[placeholderTypeId][jsonAttributeValues] = disapprovalShortNames ;
}
```
This would also simplify some of the code later on (not having to look up the index of the json, and then read that index from the disapproval names)

File: Reporting/ad-extensions-reporter.js [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+    var duplicateAttributeValuesCounts = {};
+
+    while (reportRows.hasNext()) {
+        var row = reportRows.next();
+
+        var placeholderTypeId = row['PlaceholderType']
+        if (extensionPlaceholderTypeIds.indexOf(placeholderTypeId) === -1) {
+            continue;
+        }
+        if (Object.keys(jsonAttributeValuesByExtensionType).indexOf(placeholderTypeId) === -1) {
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: To check if a javascript object has a key you can do `if("mykey" in myObject)` instead of using index of - might look nicer? Up to you which you prefer

File: Reporting/ad-extensions-reporter.js [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+
+    // Process report data, to prepare for outputting to the spreadsheet
+    Logger.log("Processing report data");
+    var extensionPlaceholderTypeIds = Object.keys(extensionPlaceholderTypeIdToFieldIdsObjectMapping);
+
+    var extensionTypeFeedItemCounts = {};
+    for (var i = 0, n = extensionPlaceholderTypeIds.length; i < n; i++) {
+        extensionTypeFeedItemCounts[extensionPlaceholderTypeIds[i]] = 0;
+    }
+    var jsonAttributeValuesByExtensionType = {};
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: Your variable names are very descriptive, which is helpful because I know what is in them by reading the name, but also makes the code a bit harder to read sometimes, as they make the lines really long, and makes it hard for me to consistently recognise the variable name (as it is hard to remember the whole thing!)

If you are using a variable a lot, sometimes it is better to have a shorter name that is easier for a human to remember, and recognise throughout the code. e.g. `extensions` or `extensionsByType`?

**Review by emma-brainlabs — COMMENTED** [review]
Nice neat code - very well laid out, good comments especially at the start.
I've just left some readability suggestions - shorten lines and eliminating some code :)

**Comment by emma-brainlabs** [comment]
Also really like the logging - giving people confidence it is doing stuff, and how you've used it to break up the code sections!

File: Reporting/ad-extensions-reporter.js (hunk 1 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yes that's better!

File: Reporting/ad-extensions-reporter.js (hunk 2 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Good point and really well explained, thanks! I like your recommendation, then we don't need to do the try catch at all!

File: Reporting/ad-extensions-reporter.js (hunk 3 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: I agree, nice!

File: Reporting/ad-extensions-reporter.js (hunk 5 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yes it does. I agree this should be clearer and I've added a comment. After sorting, `'name'` ends up at the end of the array, which is why `.pop()` works. I've made it more explicit now by using `.splice()` instead.

File: Reporting/ad-extensions-reporter.js (hunk 6 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yep, nice!

File: Reporting/ad-extensions-reporter.js (hunk 7 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js:276]: It sorts the extension rows by Feed Item Count, before outputting to the sheet. That is the penultimate value in each row (added in the line above `rowValues.push(feedItemCount, disapprovalShortNames.join(','));`).
I've added a comment.

File: Reporting/ad-extensions-reporter.js (hunk 8 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yes!

File: Reporting/ad-extensions-reporter.js (hunk 9 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Great idea, I like this much more now! I've also now included what was `duplicateAttributeValuesCounts` into this new object.

File: Reporting/ad-extensions-reporter.js (hunk 10 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Oh yes, that's nicer!

File: Reporting/ad-extensions-reporter.js (hunk 11 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yes that makes a lot of sense! It's also hard for me to remember what each variable is, and I think that slowed me down when writing the script. I've shortened some variable names.

File: Reporting/ad-extensions-reporter.js [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+    } else {
+        if (includeExtensionsWithZeroImpressions === true) {
+            var impressionsCondition = '';
+        } else if (includeExtensionsWithZeroImpressions === false) {
+            var impressionsCondition = 'WHERE Impressions > 0 ';
+        } else {
+            throw ("Value of options variable includeExtensionsWithZeroImpressions must be a boolean, either true or false. " +
+                "Currently, it is: " + includeExtensionsWithZeroImpressions)
+        }
+    }
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: This could be a function - `getZeroImpressionCondition(includeExtensionsWithZeroImpressions)`

File: Reporting/ad-extensions-reporter.js [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+
+    var placeholderFeedItemReport = AdsApp.report(
+        'SELECT PlaceholderType, AttributeValues, DisapprovalShortNames ' +
+        'FROM PLACEHOLDER_FEED_ITEM_REPORT ' +
+        impressionsCondition +
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js:218]: could even put it here
```suggestion
        getZeroImpressionsCondition(includeExtensionsWithZeroImpressions) +
```

File: Reporting/ad-extensions-reporter.js [hunk 14]
```diff
@@ -0,0 +1,384 @@
...
+            extensionsByType[placeholderTypeId] = {};
+        }
+
+        var jsonAttributes = row['AttributeValues'];
+        var jsonDisapprovalNames = row['DisapprovalShortNames'];
+        if ('' === jsonDisapprovalNames.toString()) {
+            var disapprovalNames = [];
+        } else {
+            var disapprovalNames = JSON.parse(jsonDisapprovalNames);
+        };
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: function?

File: Reporting/ad-extensions-reporter.js [hunk 15]
```diff
@@ -0,0 +1,384 @@
...
+        } else {
+            extensionsByType[placeholderTypeId][jsonAttributes]['FeedItemCount'] += 1;
+            for (var i = 0, n = disapprovalNames.length; i < n; i++) {
+                var disapprovalName = disapprovalNames[i];
+                var previouslyFoundDisapprovalNames = extensionsByType[placeholderTypeId][jsonAttributes]['Disapprovals'];
+                if (previouslyFoundDisapprovalNames.indexOf(disapprovalName) === -1) {
+                    extensionsByType[placeholderTypeId][jsonAttributes]['Disapprovals'] =
+                        previouslyFoundDisapprovalNames.concat(disapprovalName);
+                }
+            }
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: function?

File: Reporting/ad-extensions-reporter.js [hunk 16]
```diff
@@ -0,0 +1,384 @@
...
+            var extensionAttributes = JSON.parse(extension);
+            var rowValues = orderedAttributeIds.map(
+                function (attributeId) {
+                    if (extensionAttributes.hasOwnProperty(attributeId)) {
+                        return extensionAttributes[attributeId].toString();
+                    } else {
+                        return '';
+                    }
+                }
+            );
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: function?

File: Reporting/ad-extensions-reporter.js [hunk 17]
```diff
@@ -0,0 +1,384 @@
...
+        }
+    }
+
+    // Create a spreadsheet tab for each extension type, and output its data
+    Logger.log("Outputting data to spreadsheet");
+    for (var placeholderTypeId in extensionsByType) {
+        var placeholderTable = extensionTypeToPlaceholderTables[placeholderTypeId];
+        var orderedAttributeIds = Object.keys(placeholderTable).sort();
+        var nameIndex = orderedAttributeIds.indexOf('name');
+        orderedAttributeIds.splice(nameIndex, 1); // Remove 'name'
```
Comment (emma-brainlabs) [inline Reporting/ad-extensions-reporter.js]: function?

**Review by emma-brainlabs — APPROVED** [review]
All the improvements you made look great, and I agree it is easier to understand now :)

**Functions:**
Putting functions in your code is always a balancing act - too few and it is hard to follow the main flow of the code because you have to read too much nitty gritty detail. Too many and you can't understand the flow of the code _without_ going into all the functions to figure out what they are doing!

I definitely prefer code that errs of the side of too few functions, but I know other people have differing opinions!

I have tried reading the code and noting any small bits of logic that I feel are a bit admin-y (more rearranging than doing anything exciting), or just broke up what I was reading. I tried them out by deleting that code and trying to come up with a function name. If I could come up with one which wasn't too long, and felt quite clear, then I added a comment on the review. If I couldn't think of a good name, then I concluded that it probably therefore shouldn't be a function!

e.g. the zero impressions thing - reading the function name `getZeroImpressionsCondition` and seeing that it is put directly into the query, and seeing it take in the `includeExtensionsWithZeroImpressions`, tells me everything I need to know really (that it takes that user input and makes the where clause for the query) - I don't need to understand the insides of the function to understand how the code works.

Feel free to do as many or as few of these as you want!

File: Reporting/ad-extensions-reporter.js (hunk 17 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yes! It's nice that then we don't need the comment explaining what we're doing, because the function name documents it.

File: Reporting/ad-extensions-reporter.js (hunk 16 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yes!

File: Reporting/ad-extensions-reporter.js (hunk 15 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yes! And here it also simplifies the code naming-wise, because we pass in an object and because it's passed by reference we can call it whatever we like in the function itself.

File: Reporting/ad-extensions-reporter.js (hunk 14 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yes!

File: Reporting/ad-extensions-reporter.js (hunk 13 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js:218]: Very nice idea! I couldn't get it to work simply though, with having the check outside the function for whether `includeExtensionsWithZeroImpressions` is declared, so left this bit how it was.

File: Reporting/ad-extensions-reporter.js (hunk 12 above)
Comment (gideon-brainlabs) [reply Reporting/ad-extensions-reporter.js]: Yes! If `includeExtensionsWithZeroImpressions` is undeclared, I don't think we can pass it in to the function, so I've left the check for that out of the function.

**Comment by gideon-brainlabs** [comment]
Thanks for the re-review! Makes a lot of sense, I'm pleased having extracted some functions that it makes the code more readable. I've made all the functions you suggested, as well as a couple more!

**Comment by emma-brainlabs** [comment]
Looks great, I don't have any more suggestions!
